prophet/src/lib.rs

```diff
@@ -1,8 +1,153 @@
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        let result = 2 + 2;
-        assert_eq!(result, 4);
+//! A library for cloning microservice git repositories and
+use std::path::{Path, PathBuf};
+
+use git2::Repository;
+use serde::Deserialize;
+use source_code_parser::Directory;
+
+/// A cloned microservice or microservice system repository
+#[derive(Debug, Default, Deserialize)]
+pub struct MicroservicesRepository {
+    /// The Git URL of the repository to clone from
+    pub git_url: String,
+    /// The root directories to include source code files from in static analysis,
+    /// relative to `clone_dir`
+    pub root_dirs: Vec<PathBuf>,
+    /// The local directory the repository was cloned to
+    pub clone_dir: PathBuf,
+}
+
+impl MicroservicesRepository {
+    /// Clones a microservice(s) repository
+    pub fn clone(&mut self) -> Result<(), git2::Error> {
+        Repository::clone(&self.git_url, &self.clone_dir)?;
+        Ok(())
+    }
+}
+
+impl Drop for MicroservicesRepository {
+    /// Clean the cloned repository when freeing the repository from memory
+    fn drop(&mut self) {
+        if let Err(err) = std::fs::remove_dir_all(&self.clone_dir) {
+            tracing::warn!(
+                "Failed to remove cloned repository '{}' at '{:?}': {:?}",
+                self.git_url,
+                self.clone_dir,
+                err
+            );
+        }
+    }
+}
+
+/// Gets the subdirectories and files from a directory `&Path`
+fn get_dir_contents(root_dir: &Path) -> Result<(Vec<PathBuf>, Vec<PathBuf>), std::io::Error> {
+    let read_dir = match std::fs::read_dir(&root_dir) {
+        Ok(dir) => dir,
+        Err(err) => {
+            tracing::warn!("Could not read directory: {:?}", err);
+            return Err(err);
+        }
+    };
+
+    let mut files = vec![];
+    let mut sub_dirs = vec![];
+
+    for entry in read_dir {
+        let entry = entry?;
+        let entry_path = entry.path();
+
+        if entry_path.is_dir() {
+            sub_dirs.push(entry_path);
+            continue;
+        }
+
+        files.push(entry_path);
+    }
+
+    Ok((files, sub_dirs))
+}
+
+/// Convers the given root directory into its Directory representation
+fn convert_sub_dir(root_dir: PathBuf) -> Option<Directory> {
+    let (files, sub_dir_paths) = match get_dir_contents(&root_dir) {
```

Comment:
Why are we not using `?` instead of a match? Did we not want to use `anyhow`? Or can it not coerce from a Result to an Option?

Reply:
You can only use the `?` operator on the type that is the same as the return type of the function (so you can't `?` a `Result` in a function returning `Option` and vice-versa).

However, the way I did this here was a bit verbose so I simplified it to this:
```rust
let (files, sub_dir_paths) = get_dir_contents(&root_dir).map_or(None, Some)?;
```

Reply:
And `anyhow` is just for muddying Result types together so you don't need to care about specific error types, which is typically useful in application code. `anyhow` doesn't break the rule of only being able to use `?` on either `Result` or `Option` in a function, it just makes it so you can use `?` on any `Result`, since by default the `Result` also needs to return the same specific error type.

 `thiserror` is another library for helping with defining your own error types which is used in another workspace crate in this repo as well.